File: server/src/main/java/me/ajfleming/tworoomsio/service/deck/DeckBuilderService.java
package me.ajfleming.tworoomsio.service.deck;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import me.ajfleming.tworoomsio.model.Card;
import me.ajfleming.tworoomsio.model.CardSet;

public class DeckBuilderService {

  private final Map<CardSet, List<Card>> possibleCardSets = CardSetMapProvider.cardMap();

  public List<Card> buildDeck(int numberOfPlayers) {

    // Add President and Bomber
    List<Card> deck = new ArrayList<>(possibleCardSets.get(CardSet.CORE));

    // If Odd number of players, add Gambler
    addGamblerIfOddNumberOfPlayers(deck, numberOfPlayers);

    // Fill in remaining players with normal cards
    fillDeckWithMemberCards(deck, numberOfPlayers);

    return deck;
  }

  private void addGamblerIfOddNumberOfPlayers(final List<Card> deck,
      final int numberOfPlayers) {
    if (numberOfPlayers % 2 != 0) {
      deck.addAll(possibleCardSets.get(CardSet.GAMBLER));
    }
  }

  private void fillDeckWithMemberCards(final List<Card> deck, final int numberOfPlayers) {
    var cardsRequired = numberOfPlayers - deck.size();

    while (cardsRequired > 1) {
      deck.addAll(possibleCardSets.get(CardSet.FILLER));
      cardsRequired = cardsRequired - 2;
    }
  }
}

File: kotlin_server/src/main/kotlin/me/ajfleming/tworoomsio/service/DeckBuilderService.kt
package me.ajfleming.tworoomsio.service

import me.ajfleming.tworoomsio.model.Card
import me.ajfleming.tworoomsio.model.CardSet
import org.springframework.stereotype.Component

@Component
class DeckBuilderService (
    val possibleCardSets: Map<CardSet, List<Card>> = cardSetMap
) {

    fun buildDeck(numberOfPlayers: Int) : List<Card> {
        val deck : MutableList<Card> = mutableListOf()
        // Add President and Bomber
        possibleCardSets[CardSet.CORE]?.let { deck.addAll(it) }

        // If Odd number of players, add Gambler
        addGamblerIfOddNumberOfPlayers(deck, numberOfPlayers)

        // Fill in remaining players with normal cards
        fillDeckWithMemberCards(deck, numberOfPlayers)

        return deck
    }

    fun addGamblerIfOddNumberOfPlayers(deck: MutableList<Card>, numberOfPlayers: Int) {
       if (numberOfPlayers % 2 == 0) {
           possibleCardSets[CardSet.GAMBLER]?.let { deck.addAll(it) }
       }
    }

    fun fillDeckWithMemberCards(deck: MutableList<Card>, numberOfPlayers: Int) {
        var cardsRequired = numberOfPlayers - deck.size

        while(cardsRequired > 1) {
            possibleCardSets[CardSet.FILLER]?.let { deck.addAll(it) }
            cardsRequired =- 2
        }
    }
}